File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

// System.out.println(maxX);
class Solution {
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        LinkedList<Integer[]> queue = new LinkedList<>();
        int maxY = 0;
        int maxX = 0;
        int[] moveY = {0,0,1,-1};
        int[] moveX = {1,-1,0,0};
        
        for(int i = 0; i < rectangle.length; i++) {
            maxY = maxY <= rectangle[i][3] ? rectangle[i][3] : maxY;
            maxX = maxX <= rectangle[i][2] ? rectangle[i][2] : maxX;
        }
        Integer[] startPoint = {characterY*2, characterX*2, 0};
        boolean[][] visit = new boolean[maxY*2+1][maxX*2+1];
        queue.add(startPoint);
        
        while(!queue.isEmpty()) {
            Integer[] cur = queue.poll();
            int curY = cur[0];
            int curX = cur[1];
            int curK = cur[2];
            visit[curY][curX] = true;
            if(curY == itemY*2 && curX == itemX*2) {
                return curK/2;
            }
            for(int i = 0; i < moveY.length; i++) {
                int nextY = curY + moveY[i];
                int nextX = curX + moveX[i];
                int nextK = curK + 1;
                if(isExist(nextY, nextX, rectangle) && !isDuplicate(nextY, nextX, rectangle) && !visit[nextY][nextX]) {
                    Integer[] temp = {nextY, nextX, nextK};
                    queue.add(temp);
                }
            }
        }
        return -1;
    }
    //[1,2,7,5]
    //y의 범위 1 ~ 5 i = 1, i = 3
    //x의 범위 1 ~ 7 i = 0, i = 2
    
    public boolean isExist (int y, int x, int[][] A) {
        for(int i = 0; i < A.length; i++) {
            if( ((x == A[i][0]*2 || x == A[i][2]*2) && A[i][1]*2 <= y && y <= A[i][3]*2) || 
               ((y == A[i][1]*2 || y == A[i][3]*2) && A[i][0]*2 <= x && x <= A[i][2]*2))
            {
                return true;
            }
        }
        return false;
    }
    
    public boolean isDuplicate (int y, int x, int[][] A) {
        for(int i = 0; i < A.length; i++) {
            if(A[i][1]*2 < y && y < A[i][3]*2 && A[i][0]*2 < x && x < A[i][2]*2) {
                return true;
            }
        }
        return false;
    }
    
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
    let maxX = 0, maxY = 0;
    for(let i = 0; i < rectangle.length; i++){
        maxX < rectangle[i][2] ? maxX = rectangle[i][2] : undefined;
        maxY < rectangle[i][3] ? maxY = rectangle[i][3] : undefined;
    }
    let map = new Array(maxY*2+1);
    let v = new Array(maxY*2+1);
    for(let i = 0; i < map.length; i++){
        map[i] = new Array(maxX*2+1).fill(2)
        v[i] = new Array(maxX*2+1).fill(false);
    }
    
    for(let i = 0; i < rectangle.length; i++){
        for(let j = rectangle[i][1]*2; j <= rectangle[i][3]*2; j++) {
            for(let k = rectangle[i][0]*2; k <= rectangle[i][2]*2; k++){
                if((j == rectangle[i][1]*2 || j == rectangle[i][3]*2) || (k == rectangle[i][0]*2 || k == rectangle[i][2]*2)) {
                    if(map[j][k] != 0) {
                        map[j][k] = 1
                    }
                } else {
                    map[j][k] = 0;
                }
            }
        }
    }
    let startX = characterX*2;
    let startY = characterY*2;
    let endX = itemX*2;
    let endY = itemY*2;
    v[startY][startX] = true;
    let queue = [[startY,startX,0,v]];
    while(queue.length !== 0) {
        let [y,x,value,visited] = queue.shift();
        if(y == endY && x == endX) return value/2;
        
        if(map[y][x+1] != undefined && map[y][x+1] == 1 && visited[y][x+1] == false) {
            let copy = [...visited];
            for(let i = 0; i < visited.length; i++) {
                copy[i] = [...visited[i]];
            }
            copy[y][x+1] = true;
            queue.push([y,x+1,value+1,copy])
        }
        if(map[y+1] != undefined && map[y+1][x] == 1 && visited[y+1][x] == false) {
            let copy = [...visited];
            for(let i = 0; i < visited.length; i++) {
                copy[i] = [...visited[i]];
            }
            copy[y+1][x] = true;
            queue.push([y+1,x,value+1,copy])
        }
        if(map[y][x-1] != undefined && map[y][x-1] == 1 && visited[y][x-1] == false) {
            let copy = [...visited];
            for(let i = 0; i < visited.length; i++) {
                copy[i] = [...visited[i]];
            }
            copy[y][x-1] = true;
            queue.push([y,x-1,value+1,copy])
        }
        if(map[y-1] != undefined && map[y-1][x] == 1 && visited[y-1][x] == false) {
            let copy = [...visited];
            for(let i = 0; i < visited.length; i++) {
                copy[i] = [...visited[i]];
            }
            copy[y-1][x] = true;
            queue.push([y-1,x,value+1,copy])
        }
    }

}
